FWCore/Framework/src/ProductResolvers.cc

```diff
@@ -299,7 +300,8 @@ namespace edm {
                   //another thread could have finished this while we were waiting
                   setProduct(reader->getProduct(branchDescription().branchID(), &principal, mcc));
                 } catch (cms::Exception& e) {
-                  throw extendException(e, branchDescription(), mcc);
+                  extendException(e, branchDescription(), mcc);
+                  throw;
                 } catch (std::exception const& e) {
                   auto newExcept = edm::Exception(errors::StdException) << e.what();
                   throw extendException(newExcept, branchDescription(), mcc);
```

Comment:
There is one more (although a bit different) here
```suggestion
                  extendException(newExcept, branchDescription(), mcc);
                  throw nexExcept;
```

How about removing the `return` from `extendException()` function to prevent this way of slicing the exception?

Reply:
Done. I changed the function to return void and used your suggested change. This part is not a bug and works, but I agree making the function return void makes it less likely to be used incorrectly in the future. I'll push the change soon (just checking that it builds and passes unit tests at the moment).

Reply:
I'd expect this part to slice the `newExcept`, that is constructed as `edm::Exception`, to `cms::Exception`. Do I miss something?

Reply:
After looking again, I realize you are correct and I was wrong. That part was also a bug. I guess I was thrown off by the fact that it is not slicing the caught exception, but it still slices newExcept based on the function return type.